client/server/lib/logger/test/index.js

```diff
@@ -1,38 +1,78 @@
 /**
- * Internal dependencies
+ * External dependencies
  */
-import { getLogger } from '../index';
-import bunyan from 'bunyan';
+import { mockProcessStdout } from 'jest-mock-process';
+import mockFs from 'mock-fs';
+import fs from 'fs';
 
-/**
- * Mocks `bunyan` so every instance returned by `createLogger` logs to a buffer instead
- * of process.stdout. We can retrieve that buffer with `bunyan.getOutputBuffer()` and make
- * assertions on it.
- */
-jest.mock( 'bunyan', () => {
-	const realBunyan = jest.requireActual( 'bunyan' );
-	const buffer = new realBunyan.RingBuffer( { limit: 100 } );
-
-	return {
-		createLogger: jest.fn( ( options ) => {
-			return realBunyan.createLogger( {
-				...options,
-				streams: [ { type: 'raw', stream: buffer } ],
-			} );
-		} ),
-		getOutputBuffer: () => buffer,
-	};
+let mockStdout;
+let getLogger;
+
+beforeEach( () => {
+	( { getLogger } = require( '../index' ) );
+	mockStdout = mockProcessStdout();
+	mockFs( {
+		'/tmp': {},
+	} );
+} );
+
+afterEach( () => {
+	jest.resetModules();
+	mockStdout.mockRestore();
+	delete process.env.CALYPSO_LOGFILE;
+	mockFs.restore();
 } );
 
 it( "Returns a logger with the name 'calypso'", () => {
-	const buffer = bunyan.getOutputBuffer();
+	const logger = getLogger();
+
+	logger.info( 'message' );
+
+	expect( mockStdout ).toHaveBeenCalledTimes( 1 );
+	const loggedMessage = mockStdout.mock.calls[ 0 ][ 0 ];
+	expect( JSON.parse( loggedMessage ) ).toEqual( expect.objectContaining( { name: 'calypso' } ) );
+} );
+
+it( 'Logs info and above levels to stdout', () => {
+	const logger = getLogger();
+
+	logger.trace( 'trace' ); // not logged
+	logger.debug( 'debug' ); // not logged
+	logger.info( 'info' );
+	logger.warn( 'warn' );
+	logger.error( 'error' );
+	logger.fatal( 'fatal' );
+
+	const loggedMessages = mockStdout.mock.calls.map( ( args ) => JSON.parse( args[ 0 ] ) );
+	expect( loggedMessages ).toEqual( [
+		expect.objectContaining( { msg: 'info', level: 30 } ),
+		expect.objectContaining( { msg: 'warn', level: 40 } ),
+		expect.objectContaining( { msg: 'error', level: 50 } ),
+		expect.objectContaining( { msg: 'fatal', level: 60 } ),
+	] );
+} );
+
+it( 'Logs info and above levels to the filesystem when the env variable is present', async () => {
+	process.env.CALYPSO_LOGFILE = '/tmp/calypso.log';
+	const logger = getLogger();
```

Comment:
If the `createLogger` had an argument called `logFile`, and if `getLogger` passed `process.env.CALYPSO_LOGFILE` as the argument, and if you used `createLogger` in the test, then the module would become slightly more unit-testable: instead of modifying the environment, we would simply pass arguments to functions.

Using `createLogger` in most unit tests would be preferrable: we want to create a new logger for each unit test. Without relying on `jest.resetModules` to reset the singleton.

The only place for `getLogger` is actually testing whether the singleton is initialized exactly once.

Reply:
I would like to avoid exposing `createLogger()`, as nobody should be calling that function. If they want their own custom logger they should be use `getLogger().child()`, the style [proposed by bunyan](https://github.com/trentm/node-bunyan#logchild). That way we can maintain control of streams, levels, mandatory fields, etc.

When writing other unit test that use the logger, they should mock `getLogger()` and assert the calls they make to the logger library, not the actual logging behaviour (otherwise it won't be an unit test).

I prefer to wait to see how this logger gets used across the codebase and refactor as necessary